ramips: add support for Xiaomi MiWifi 3C

eduardo010174 opened the pull request:
This commit adds support for Xiaomi MiWiFi 3C device.

Xiaomi MiWifi 3C has almost the same system architecture as the Xiaomi Mi WiFi Nano, which is already officially supported by OpenWrt.

The differences are:

 - Numbers of antennas (4 instead of 2). The antenna management is done via the µC. There is no configuration needed in the software code.
 - LAN port assignments are different. LAN1 and WAN are interchanged.

OpenWrt Wiki: https://openwrt.org/toh/xiaomi/mir3c

OpenWrt developers forum page: https://forum.openwrt.org/t/support-for-xiaomi-mi-3c

Specifications:

 - CPU: MediaTek MT7628AN (575MHz)
 - Flash: 16MB
 - RAM: 64MB DDR2
 - 2.4 GHz: IEEE 802.11b/g/n with Integrated LNA and PA
 - Antennas: 4x external single band antennas
 - WAN: 1x 10/100M
 - LAN: 2x 10/100M
 - LED: 1x amber/blue/red. Programmable
 - Button: Reset

MAC addresses as verified by OEM firmware:

use address source
LAN *:92 factory 0x28
WAN *:92 factory 0x28
2g *:93 factory 0x4

OEM firmware uses VLAN's to create the network interface for WAN and LAN.

Bootloader info:
The stock bootloader uses a "Dual ROM Partition System". OS1 is a deep copy of OS2.
The bootloader start OS2 by default.
To force start OS1 it is needed to set "flag_try_sys2_failed=1".

How to install:
1- Use OpenWRTInvasion to gain telnet, ssh and ftp access.
https://github.com/acecilia/OpenWRTInvasion
(IP: 192.168.31.1 - Username: root - Password: root)
2- Connect to router using telnet or ssh.
3- Backup all partitions. Use command  "dd if=/dev/mtd0 of=/tmp/mtd0". Copy /tmp/mtd0 to computer using ftp.
4- Copy openwrt-ramips-mt76x8-xiaomi_miwifi-3c-squashfs-sysupgrade.bin to /tmp in router using ftp.
5- Enable UART access and change start image for OS1.
```
nvram set uart_en=1
nvram set flag_last_success=1
nvram set boot_wait=on
nvram set flag_try_sys2_failed=1
nvram commit
```
6- Installing Openwrt on OS1 and free OS2.
```
mtd erase OS1
mtd erase OS2
mtd -r write /tmp/openwrt-ramips-mt76x8-xiaomi_miwifi-3c-squashfs-sysupgrade.bin OS1
```

Limitations: For the first install the image size needs to be less than 7733248 bits.

Thanks for all community and especially for this device:
minax007, earth08, S.Farid

Signed-off-by: Eduardo Santos <edu.2000.kill@gmail.com>

adschm commented:
Please don't double-post, but push your changes to the existing PR next time.

eduardo010174 commented:
> Please don't double-post, but push your changes to the existing PR next time.

Sorry. It's my first pull request. I thought it would be an error for not having created a branch other than the master.

adschm commented:
Replacing bootloader is typically a no-go on ramips. Let's see what other reviewers think ...

adschm commented on target/linux/ramips/dts/mt7628an_xiaomi_miwifi-3c.dts:
```diff
@@ -0,0 +1,120 @@
+//SPDX-License-Identifier: GPL-2.0-or-later OR MIT
```
Missing space after "//"

eduardo010174 replied:
Fixed 3c02f37.

adschm commented on target/linux/ramips/dts/mt7628an_xiaomi_miwifi-3c.dts:
```diff
@@ -0,0 +1,120 @@
+};
+
+&esw {
+	mediatek,portmap = <0x2f>;
```
This would correspond to 4=wan in switch setup?

eduardo010174 replied:
I looked for documentation the only reference I found was this one. https://patchwork.ozlabs.org/project/lede/patch/20170222143634.GA3820@makrotopia.org/. I change portmap for "llllw". But for portdisable I didn't find anything. 9a8c702

adschm replied:
For portmap, 0x2f is binary 101111. The zero marks the wan port. So, just need to adjust that so 0 becomes the real wan port, which is 0 based on your configuration (assuming that your configuration is correct).

port-disable seemed correct at first check. It's the same concept, where you will have binary 1 for disable and 0 for enable ...

adschm replied:
"llllw" is for other platforms (mt7620) ...

eduardo010174 replied:
02_network that is correct and tested.

For portmap, 011111 -> 0x1f
For portdisable, 010101 -> 0x15

It's correct? @adschm

adschm replied:
Err, no. Would you write "one thousand" as 0001?

eduardo010174 replied:
I write 1000 ahhaha.
Ok. MSB <- LSB.
I need flip bits.

For portmap, 111110 -> 0x3e
For portdisable, 101010 -> 0x2a

It's correct? @adschm

eduardo010174 replied:
Fixed 9a8c702

adschm replied:
Well, if you resolve it, how should I read the answer?

adschm replied:
Seems correct now, though.

adschm commented on target/linux/ramips/dts/mt7628an_xiaomi_miwifi-3c.dts:
```diff
@@ -0,0 +1,120 @@
+		};
+	};
+};
+
```
remove empty line at EOF.

eduardo010174 replied:
Fixed 659e2ce.

adschm commented on target/linux/ramips/mt76x8/base-files/etc/board.d/02_network:
```diff
@@ -247,6 +251,7 @@ ramips_setup_macs()
 	xiaomi,mi-router-4c)
 		wan_mac=$(mtd_get_mac_binary factory 0x4)
 		;;
+	xiaomi,miwifi-3c|\
```
What?

eduardo010174 replied:
It's not needed. Fixed 7e7c731.

adschm commented on target/linux/ramips/mt76x8/base-files/etc/init.d/bootcount:
```diff
@@ -9,6 +9,9 @@ boot() {
 			echo -e "bootcount\nbootchanged\n" | /usr/sbin/fw_setenv -s -
 		;;
 	xiaomi,mi-router-4c|\
+	xiaomi,miwifi-3c)
```
Err, why separate the -nano?

eduardo010174 replied:
Is not needed in breed bootloader. Fixed 2c753b0.

adschm reviewed (changes requested):
The commit message should be a _commit_ message. Please also add a MAC address overview like here (if you can still find out about the OEM MAC address assignment):

https://openwrt.org/docs/guide-developer/device-support-policies#mac_addresses

eduardo010174 commented:
@minax007 test build with all changes and not report regressions. [https://forum.openwrt.org/t/support-for-xiaomi-mi-3c/84973/339](url)

adschm commented:
Still, no commit message, no MAC addresses added to the commit message, and no proper commit title....

eduardo010174 commented:
I and @minax007 working now in MAC addresses.
"Commit message" is need add to real commit? I didn't understand this part of the commit  and "no proper commit title".

adschm commented:
_Commit_ message means it's the message of the commit. Not the initial post of the PR.

The commit title has certain requirements. Just go to the commit history and have a look ...

eduardo010174 commented:
> _Commit_ message means it's the message of the commit. Not the initial post of the PR.
> 
> The commit title has certain requirements. Just go to the commit history and have a look ...

Ok added Macs Layout and fixes in commit message and title.

adschm commented:
> eth0     *:92       factory 0x28
> 2g        *:93       factory 0x4

What about WAN?

eduardo010174 commented:
> > eth0     *:92       factory 0x28
> > 2g        *:93       factory 0x4
> 
> What about WAN?

The stock firmware use for LAN eth0.1 with same mac for WAN eth0.2 .

eduardo010174 commented:
> The commit message should be a _commit_ message. Please also add a MAC address overview like here (if you can still find out about the OEM MAC address assignment):
> 
> https://openwrt.org/docs/guide-developer/device-support-policies#mac_addresses

Fixed 79b7f32

adschm commented:
Replacing the bootloader is typically a no-go on this target ...

Neustradamus commented:
To follow this PR.

eduardo010174 committed:
ramips: add support for Xiaomi MiWifi 3C

This commit adds support for Xiaomi MiWiFi 3C device.

Xiaomi MiWifi 3C has almost the same system architecture as the Xiaomi Mi WiFi Nano, which is already officially supported by OpenWrt.

The differences are:

 - Numbers of antennas (4 instead of 2). The antenna management is done via the µC. There is no configuration needed in the software code.
 - LAN port assignments are different. LAN1 and WAN are interchanged.

OpenWrt Wiki: https://openwrt.org/toh/xiaomi/mir3c

OpenWrt developers forum page: https://forum.openwrt.org/t/support-for-xiaomi-mi-3c

Specifications:

 - CPU: MediaTek MT7628AN (575MHz)
 - Flash: 16MB
 - RAM: 64MB DDR2
 - 2.4 GHz: IEEE 802.11b/g/n with Integrated LNA and PA
 - Antennas: 4x external single band antennas
 - WAN: 1x 10/100M
 - LAN: 2x 10/100M
 - LED: 1x amber/blue/red. Programmable
 - Button: Reset

MAC addresses as verified by OEM firmware:

use address source
LAN *:92 factory 0x28
WAN *:92 factory 0x28
2g *:93 factory 0x4

OEM firmware uses VLAN's to create the network interface for WAN and LAN.

Bootloader info:
The stock bootloader uses a "Dual ROM Partition System". OS1 is a deep copy of OS2.
The bootloader start OS2 by default.
To force start OS1 it is needed to set "flag_try_sys2_failed=1".

How to install:
1- Use OpenWRTInvasion to gain telnet, ssh and ftp access.
https://github.com/acecilia/OpenWRTInvasion
(IP: 192.168.31.1 - Username: root - Password: root)
2- Connect to router using telnet or ssh.
3- Backup all partitions. Use command  "dd if=/dev/mtd0 of=/tmp/mtd0". Copy /tmp/mtd0 to computer using ftp.
4- Copy openwrt-ramips-mt76x8-xiaomi_miwifi-3c-squashfs-sysupgrade.bin to /tmp in router using ftp.
5- Enable UART access and change start image for OS1.
```
nvram set uart_en=1
nvram set flag_last_success=1
nvram set boot_wait=on
nvram set flag_try_sys2_failed=1
nvram commit
```
6- Installing Openwrt on OS1 and free OS2.
```
mtd erase OS1
mtd erase OS2
mtd -r write /tmp/openwrt-ramips-mt76x8-xiaomi_miwifi-3c-squashfs-sysupgrade.bin OS1
```

Limitations: For the first install the image size needs to be less than 7733248 bits.

Thanks for all community and especially for this device:
minax007, earth08, S.Farid

Signed-off-by: Eduardo Santos <edu.2000.kill@gmail.com>

adschm commented:
Thanks, will merge it with a few adjustments.

Neustradamus commented:
@adschm: I do not find the commit, where is it?

Neustradamus commented:
Merged commit is:
- https://github.com/openwrt/openwrt/commit/3c97fb4346d11ab3e7cc453ffd7b66b4b9140aec

Thanks!
